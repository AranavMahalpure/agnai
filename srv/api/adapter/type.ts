import { AppSchema } from '../../db/schema'

export type ModelAdapter = (opts: {
  chat: AppSchema.Chat
  char: AppSchema.Character
  user: AppSchema.User
  members: AppSchema.Profile[]
  message: string
  sender: AppSchema.Profile
  prompt: string
  genSettings: any
}) => AsyncGenerator<string | { error: any }>
