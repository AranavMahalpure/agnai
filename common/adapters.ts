export type AIAdapter = (typeof AI_ADAPTERS)[number]
export type ChatAdapter = (typeof CHAT_ADAPTERS)[number]
export type PersonaFormat = (typeof PERSONA_FORMATS)[number]

export const PERSONA_FORMATS = ['boostyle', 'wpp', 'sbf'] as const

export const PERSONA_LABELS: { [key in PersonaFormat]: string } = {
  boostyle: 'Boostyle',
  wpp: 'W++',
  sbf: 'SBF',
}

export const AI_ADAPTERS = ['kobold', 'novel', 'chai', 'ooba', 'horde'] as const
export const CHAT_ADAPTERS = ['default', ...AI_ADAPTERS] as const

export const MULTI_TENANT_ADAPTERS = ['novel', 'chai', 'kobold'] as const

export type NovelModel = keyof typeof NOVEL_MODELS

export const NOVEL_MODELS = {
  euterpe: 'euterpe-v2',
  krake: 'krake-v2',
} satisfies { [key: string]: string }

export type HordeModel = {
  name: string
  count: number
  performance: number
  queued: number
  eta: number
  type?: string
}

export const ADAPTER_LABELS: Record<AIAdapter, string> = {
  chai: 'Chai',
  horde: 'Horde',
  kobold: 'Kobold',
  novel: 'NovelAI',
  ooba: 'Text-Generation-WebUI',
}
