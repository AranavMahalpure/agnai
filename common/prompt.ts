import gpt from 'gpt-3-encoder'
import { AppSchema } from '../srv/db/schema'
import { defaultPresets } from './presets'

const DEFAULT_MAX_TOKENS = 2048

export type PromptOpts = {
  chat: AppSchema.Chat
  char: AppSchema.Character
  members: AppSchema.Profile[]
  settings?: AppSchema.GenSettings
  messages: AppSchema.ChatMessage[]
  retry?: AppSchema.ChatMessage
  continue?: string
}

export const BOT_REPLACE = /\{\{char\}\}/g
export const SELF_REPLACE = /\{\{user\}\}/g

export function createPrompt(opts: PromptOpts) {
  const { chat, char, members, settings = defaultPresets.basic, messages, retry } = opts
  const { pre, post } = createPromptSurrounds(opts)
  const maxContext = settings.maxContextLength || defaultPresets.basic.maxContextLength

  const lines = getLinesForPrompt(opts)
  const history: string[] = []
  let tokens = gpt.encode([pre, post].join('\n')).length

  for (const text of lines) {
    const size = gpt.encode(text).length

    if (size + tokens > maxContext) break

    history.push(text)
    tokens += size
  }

  const prompt = [pre, ...history, post].filter(removeEmpty).join('\n')
  return { lines, prompt }
}

function getHandle(members: AppSchema.Profile[], id?: string) {
  if (!id) return 'You'

  for (const mem of members) {
    if (mem.userId === id) return mem.handle || 'You'
  }

  return 'You'
}

export function createPromptSurrounds(
  opts: Pick<PromptOpts, 'chat' | 'char' | 'members' | 'continue'>
) {
  const { chat, char, members } = opts
  const sender = members.find((mem) => mem.userId === chat.userId)?.handle || 'You'

  const hasStartSignal =
    chat.scenario.includes('<START>') ||
    chat.sampleChat.includes('<START>') ||
    chat.greeting.includes('<START>')

  const pre: string[] = [`${char.name}'s Persona: ${formatCharacter(char.name, chat.overrides)}`]

  if (chat.scenario) pre.push(`Scenario: ${chat.scenario}`)

  if (!hasStartSignal) pre.push('<START>')

  const sampleChat = chat.sampleChat.split('\n').filter(removeEmpty)
  pre.push(...sampleChat)

  const post = [`${char.name}:`]
  if (opts.continue) {
    post.unshift(`${char.name}: ${opts.continue}`)
  }

  return {
    pre: pre.join('\n').replace(BOT_REPLACE, char.name).replace(SELF_REPLACE, sender),
    post: `${char.name}:`,
  }
}

type PromptParts = {
  scenario?: string
  greeting?: string
  sampleChat?: string[]
  persona: string
  post: string[]
}

export function getPromptParts(
  opts: Pick<PromptOpts, 'chat' | 'char' | 'members' | 'continue'>,
  personaKind?: AppSchema.CharacterPersona['kind']
) {
  const { chat, char, members } = opts
  const sender = members.find((mem) => mem.userId === chat.userId)?.handle || 'You'

  const replace = (value: string) => placeholderReplace(value, char.name, sender)

  const parts: PromptParts = {
    persona: formatCharacter(char.name, chat.overrides),
    post: [],
  }

  if (chat.scenario) {
    parts.scenario = replace(chat.scenario)
  }

  if (chat.sampleChat) {
    parts.sampleChat = chat.sampleChat.split('\n').filter(removeEmpty).map(replace)
  }

  if (chat.greeting) {
    parts.greeting = replace(chat.greeting)
  }

  const post = [`${char.name}:`]
  if (opts.continue) {
    post.unshift(`${char.name}: ${opts.continue}`)
  }

  parts.post = post.map(replace)

  return parts
}

function placeholderReplace(value: string, charName: string, senderName: string) {
  return value.replace(BOT_REPLACE, charName).replace(SELF_REPLACE, senderName)
}

export function formatCharacter(name: string, persona: AppSchema.CharacterPersona) {
  switch (persona.kind) {
    case 'wpp': {
      const attrs = Object.entries(persona.attributes)
        .map(([key, values]) => `${key}(${values.map(quote).join(' + ')})`)
        .join('\n')

      return [`[character("${name}") {`, attrs, '}]'].join('\n')
    }

    case 'sbf': {
      const attrs = Object.entries(persona.attributes).map(
        ([key, values]) => `${key}: ${values.map(quote).join(', ')}`
      )

      return `[ character: "${name}"; ${attrs.join('; ')} ]`
    }

    case 'boostyle': {
      const attrs = Object.values(persona.attributes).reduce(
        (prev, curr) => {
          prev.push(...curr)
          return prev
        },
        [name]
      )
      return attrs.join(' + ')
    }

    case 'text': {
      return persona.attributes.text[0]
    }
  }
}

export function exportCharacter(char: AppSchema.Character, target: 'tavern' | 'ooba') {
  switch (target) {
    case 'tavern': {
      return {
        name: char.name,
        first_mes: char.greeting,
        scenario: char.scenario,
        description: formatCharacter(char.name, char.persona),
        mes_example: char.sampleChat,
      }
    }

    case 'ooba': {
      return {
        char_name: char.name,
        char_greeting: char.greeting,
        world_scenario: char.scenario,
        char_persona: formatCharacter(char.name, char.persona),
        example_dialogue: char.sampleChat,
      }
    }
  }
}

function quote(str: string) {
  return `"${str}"`
}

function removeEmpty(value?: string) {
  return !!value
}

function sortDesc(left: AppSchema.ChatMessage, right: AppSchema.ChatMessage) {
  return left.createdAt > right.createdAt ? 1 : left.createdAt === right.createdAt ? 0 : -1
}

/**
 * We 'ambitiously' get enough tokens to fill up the entire prompt.
 *
 * In `createPrompt()`, we trim this down to fit into the context with all of the chat and character context
 */
export function getLinesForPrompt(
  { chat, settings, char, members, messages }: PromptOpts,
  lines: string[] = []
) {
  const maxContext = settings?.maxContextLength || DEFAULT_MAX_TOKENS
  let tokens = 0

  const formatMsg = (chat: AppSchema.ChatMessage) => prefix(chat, char.name, members) + chat.msg

  const history = messages.map(formatMsg)

  for (const hist of history) {
    const nextTokens = gpt.encode(hist).length
    if (nextTokens + tokens > maxContext) break
    tokens += nextTokens
    lines.unshift(hist)
  }

  return lines
}

function prefix(chat: AppSchema.ChatMessage, bot: string, members: AppSchema.Profile[]) {
  const member = members.find((mem) => chat.userId === mem.userId)

  return chat.characterId ? `${bot}: ` : `${member?.handle}: `
}
