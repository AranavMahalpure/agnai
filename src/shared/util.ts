import { UnwrapBody, assertValid } from 'frisker'

type FormRef = { [key: string]: 'string' | 'string?' }

export function getForm<T extends FormRef>(evt: Event, body: T) {
  evt.preventDefault()
  if (!evt.target) return
  const target = evt.target
  const form = new FormData(target as HTMLFormElement)
  const values = Object.keys(body).reduce((prev, curr) => {
    const value = form.get(curr)?.toString()
    prev[curr] = value
    return prev
  }, {} as any) as UnwrapBody<T>

  assertValid(body, values)
  return values
}
