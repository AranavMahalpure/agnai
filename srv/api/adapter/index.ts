import { store } from '../../db'
import { AppSchema } from '../../db/schema'
import { handleKobold } from './kobold'

export async function generateResponse(
  adapter: AppSchema.ChatAdapter,
  chat: AppSchema.Chat,
  char: AppSchema.Character,
  message: string,
  history: AppSchema.ChatMessage[]
) {
  if (!adapter || adapter === 'none') {
    throw new Error(`No adapter set`)
  }

  switch (adapter) {
    case 'kobold':
    default: {
      return handleKobold(chat, char, history, message)
    }
  }
}
