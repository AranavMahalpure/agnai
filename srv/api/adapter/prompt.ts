import { AppSchema } from '../../db/schema'
import gpt from 'gpt-3-encoder'
import { logger } from '../../logger'
import { store } from '../../db'

type PromptOpts = {
  sender: AppSchema.Profile
  chat: AppSchema.Chat
  char: AppSchema.Character
  message: string
  members: AppSchema.Profile[]
  retry?: AppSchema.ChatMessage
}

const MAX_TOKENS = 2048
const BOT_REPLACE = /\{\{char\}\}/g
const SELF_REPLACE = /\{\{user\}\}/g

export async function createPrompt({ sender, chat, char, message, members }: PromptOpts) {
  const username = sender.handle || 'You'

  const pre: string[] = [`${char.name}'s Persona: ${formatCharacter(char.name, chat.overrides)}`]

  if (chat.scenario) {
    pre.push(`Scenario: ${chat.scenario}`)
  }

  pre.push(`<START>`, ...chat.sampleChat.split('\n'))
  const post = [`${username}: ${message}`, `${char.name}:`]

  const prompt = await appendHistory(chat, members, char, pre, post)
  return prompt
}

async function appendHistory(
  chat: AppSchema.Chat,
  members: AppSchema.Profile[],
  char: AppSchema.Character,
  pre: string[],
  post: string[],
  retry?: AppSchema.ChatMessage
) {
  const owner = members.find((mem) => mem.userId === chat.userId)
  if (!owner) {
    throw new Error(`Cannot produce prompt: Owner profile not found`)
  }

  // We need to do this early for accurate token counts
  const preamble = pre
    .filter(removeEmpty)
    .join('\n')
    .replace(BOT_REPLACE, char.name)
    .replace(SELF_REPLACE, owner.handle)
  const postamble = post
    .filter(removeEmpty)
    .join('\n')
    .replace(BOT_REPLACE, char.name)
    .replace(SELF_REPLACE, owner.handle)

  let tokens = gpt.encode(preamble + '\n' + postamble).length
  const lines: string[] = []
  let before = retry?.updatedAt

  do {
    const messages = await store.chats.getRecentMessages(chat._id, before)
    const history = messages.map((chat) => prefix(chat, char.name, members) + chat.msg)

    for (const hist of history) {
      const nextTokens = gpt.encode(hist).length
      if (nextTokens + tokens > MAX_TOKENS) break
      tokens += nextTokens
      lines.unshift(hist)
    }

    if (tokens >= MAX_TOKENS || messages.length < 50) break
    before = messages.slice(-1)[0].createdAt
  } while (true)

  const middle = lines
    .join('\n')
    .replace(BOT_REPLACE, char.name)
    .replace(SELF_REPLACE, owner.handle)

  const prompt = [preamble, middle, postamble].join('\n')
  logger.info({ tokens, prompt }, 'Tokens used')
  return prompt
}

export function formatCharacter(name: string, persona: AppSchema.CharacterPersona) {
  switch (persona.kind) {
    case 'wpp': {
      const attrs = Object.entries(persona.attributes)
        .map(([key, values]) => `${key}(${values.map(quote).join(' + ')})`)
        .join('\n')

      return [`[character("${name}") {`, attrs, '}]'].join('\n')
    }

    case 'sbf': {
      const attrs = Object.entries(persona.attributes).map(
        ([key, values]) => `${key}: ${values.map(quote).join(', ')}`
      )

      return `[ character: "${name}"; ${attrs.join('; ')} ]`
    }

    case 'boostyle': {
      const attrs = Object.values(persona.attributes).reduce(
        (prev, curr) => {
          prev.push(...curr)
          return prev
        },
        [name]
      )
      return attrs.join(' + ')
    }
  }
}

export function exportCharacter(char: AppSchema.Character, target: 'tavern' | 'ooba') {
  switch (target) {
    case 'tavern': {
      return {
        name: char.name,
        first_mes: char.greeting,
        scenario: char.scenario,
        description: formatCharacter(char.name, char.persona),
        mes_example: char.sampleChat,
      }
    }

    case 'ooba': {
      return {
        char_name: char.name,
        char_greeting: char.greeting,
        world_scenario: char.scenario,
        char_persona: formatCharacter(char.name, char.persona),
        example_dialogue: char.sampleChat,
      }
    }
  }
}

function quote(str: string) {
  return `"${str}"`
}

function prefix(chat: AppSchema.ChatMessage, bot: string, members: AppSchema.Profile[]) {
  const member = members.find((mem) => chat.userId === mem.userId)

  return chat.characterId ? `${bot}: ` : `${member?.handle}: `
}

function removeEmpty(value?: string) {
  return !!value
}
