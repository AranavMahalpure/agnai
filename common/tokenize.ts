import { AIAdapter, OPENAI_MODELS } from './adapters'
import gpt from 'gpt-3-encoder'

export type Encoder = (value: string) => number

const main: Encoder = function main(value: string) {
  return gpt.encode(value).length
}

const novel: Encoder = function krake(value: string) {
  return gpt.encode(value).length + 4
}

let davinci: Encoder
let turbo: Encoder

export function getEncoder(adapter: AIAdapter | 'main', model?: string) {
  if (adapter !== 'openai' && adapter !== 'novel') return main

  if (adapter === 'novel') {
    return novel
  }

  if (model === OPENAI_MODELS.DaVinci) {
    return davinci ?? novel
  }

  if (model === OPENAI_MODELS.Turbo) {
    return turbo ?? novel
  }

  return main
}
