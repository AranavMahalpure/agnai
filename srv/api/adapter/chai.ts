import needle from 'needle'
import { config } from '../../config'
import { logger } from '../../logger'
import { trimResponse } from '../chat/common'
import { createPrompt } from './prompt'
import { ModelAdapter } from './type'

const base = {
  repetition_penalty: 1.1,
  response_length: 64,
  tempature: 0.6,
  top_k: 40,
  top_p: 1,
}

export const handleChai: ModelAdapter = async function* ({
  chat,
  char,
  history,
  message,
  sender,
  members,
}) {
  if (!config.chai.url) {
    yield { error: 'Chai URL not set' }
    return
  }

  if (config.chai.uid === 'empty' || config.chai.key === 'empty') {
    yield { error: 'Chai is not configured' }
    return
  }

  const body = {
    ...base,
    text: createPrompt({ sender, chat, char, history, message }),
  }

  const response = await needle('post', `${config.chai.url}/generate/gptj`, body, {
    json: true,
    headers: {
      'user-agent':
        'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36',
      origin: 'https://chai.ml',
      developer_uid: config.chai.uid,
      developer_key: config.chai.key,
    },
  })
  logger.warn(response.body, 'Chai response')

  const status = response.statusCode || 0
  if (status >= 500) {
    yield { error: 'Chai internal server error' }
    return
  }

  if (status >= 400) {
    yield { error: 'Chai request error' }
    return
  }

  const username = 'You'
  const endTokens = [
    `Me:`,
    `Me :`,
    `${username}:`,
    `${char.name}:`,
    `${username} :`,
    `${char.name} :`,
    'END_OF_DIALOG',
  ]

  const trimmed = trimResponse(response.body.data, char, members, endTokens)
  yield trimmed ? trimmed.response : response.body.data
}
