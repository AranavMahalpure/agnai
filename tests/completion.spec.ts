import './init'
import { expect } from 'chai'
import { toChar, toChat, toProfile, toUser } from './util'
import { OPENAI_MODELS } from '/common/adapters'
import { createPromptWithParts, getPromptParts } from '/common/prompt'
import { toChatCompletionPayload } from '/srv/adapter/chat-completion'
import { logger } from '/srv/logger'
import { getEncoder } from '/srv/tokenize'
import { GenerateRequestV2 } from '/srv/adapter/type'
import { defaultPresets } from '/common/default-preset'
import { AppSchema } from '/common/types'

const settings: Partial<AppSchema.GenSettings> = {
  oaiModel: OPENAI_MODELS.Turbo,
  service: 'openai',
  maxContextLength: 4096,
  maxTokens: 128,
  systemPrompt: 'This is a system prompt',
  useGaslight: true,
  gaslight: defaultPresets.openai.gaslight,
}

const user = toUser('Admin')
const profile = toProfile('Oreo')
const char = toChar('Robot', {
  scenario: 'Snapshot testing scenario',
  sampleChat: `{{char}}: This is how I speak\nOreo: I speak to you like this`,
})
const chat = toChat(char, {})
const characters = { [char._id]: char }
const lines = [`Robot: Greetings!`, 'Oreo: My response to your greeting!']

const encoder = () => getEncoder('openai', OPENAI_MODELS.Turbo)

describe('Chat completion', () => {
  it('will', () => {
    const actual = getPayload()
    expect(actual).toMatchSnapshot()
  })
})

type TestOpts = {
  settings?: Partial<AppSchema.GenSettings>
  char?: AppSchema.Character
  lines?: string[]
}

function getPayload(test: TestOpts = {}) {
  const { parts, opts, lines } = getOpts(test)

  const prompt = createPromptWithParts(opts, parts, lines, encoder())

  const payload = toChatCompletionPayload(
    {
      requestId: '',
      char: test.char || char,
      chat,
      gen: Object.assign({}, opts.settings, test.settings),
      kind: 'send',
      replyAs: char,
      log: logger,
      lines: prompt.lines,
      members: [profile],
      parts: prompt.parts,
      prompt: prompt.prompt,
      characters,
      impersonate: undefined,
      sender: profile,
      settings: {},
      user,
    },
    1024
  )

  return payload
}

function getOpts(test: TestOpts) {
  const history = test.lines || lines
  const parts = getPromptParts(
    {
      char,
      chat,
      members: [profile],
      replyAs: char,
      user,
      characters,
      settings,
      kind: 'send',
    },
    history,
    encoder()
  )

  const opts: GenerateRequestV2 = {
    requestId: '',
    char,
    chat,
    kind: 'send' as const,
    characters,
    lines: history,
    members: [profile],
    parts,
    replyAs: char,
    sender: profile,
    user,
    settings,
  }

  return { parts, opts, lines: history }
}
