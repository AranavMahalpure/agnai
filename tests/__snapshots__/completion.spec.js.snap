// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat completion will 1`] = `
Array [
  Object {
    "content": "
This is a system prompt
Description of Robot:
PERSONA
Circumstances and context of the dialogue: Snapshot testing scenario",
    "role": "system",
  },
  Object {
    "content": "How Robot speaks:",
    "role": "system",
  },
  Object {
    "content": "Robot: This is how I speak",
    "role": "example-assistant",
  },
  Object {
    "content": "Oreo: I speak to you like this
System: New conversation started. Previous conversations are examples only.",
    "role": "example-user",
  },
  Object {
    "content": "Robot: Greetings!",
    "role": "assistant",
  },
  Object {
    "content": "Oreo: My response to your greeting!",
    "role": "user",
  },
  Object {
    "content": "Respond as Robot",
    "role": "system",
  },
]
`;
